File: flutter/ios/Runner/Channels/SocketChannelHandler.swift
import Foundation
import Flutter
import SocketIO

class SocketChannelHandler: NSObject, FlutterStreamHandler {
    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?
    private var manager: SocketManager?
    private var socket: SocketIOClient?
    private var isConnected = false

    init(messenger: FlutterBinaryMessenger) {
        methodChannel = FlutterMethodChannel(
            name: "com.callsafe.socket",
            binaryMessenger: messenger
        )
        eventChannel = FlutterEventChannel(
            name: "com.callsafe.socket.events",
            binaryMessenger: messenger
        )
        super.init()

        methodChannel.setMethodCallHandler(handle)
        eventChannel.setStreamHandler(self)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "connect":
            guard let args = call.arguments as? [String: Any],
                  let token = args["token"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing token", details: nil))
                return
            }
            let serverUrl = args["serverUrl"] as? String ?? "wss://tunnel.callsafe.tech"
            connect(token: token, serverUrl: serverUrl, result: result)

        case "disconnect":
            disconnect()
            result(nil)

        case "emit":
            guard let args = call.arguments as? [String: Any],
                  let event = args["event"] as? String,
                  let data = args["data"] as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            emit(event: event, data: data)
            result(nil)

        case "isConnected":
            result(isConnected)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func connect(token: String, serverUrl: String, result: @escaping FlutterResult) {
        print("[Socket] Connecting to: \(serverUrl)")

        // Disconnect existing connection
        disconnect()

        guard let url = URL(string: serverUrl) else {
            result(FlutterError(code: "INVALID_URL", message: "Invalid server URL", details: nil))
            return
        }

        let config: SocketIOClientConfiguration = [
            .log(false),
            .compress,
            .forceWebsockets(true),
            .connectParams(["token": token]),
            .version(.three)
        ]

        manager = SocketManager(socketURL: url, config: config)
        socket = manager?.defaultSocket

        setupSocketHandlers()
        socket?.connect()
        result(nil)
    }

    private func setupSocketHandlers() {
        socket?.on(clientEvent: .connect) { [weak self] _, _ in
            print("[Socket] Connected")
            self?.isConnected = true
            self?.eventSink?([
                "type": "connected"
            ])
            self?.registerDevice()
        }

        socket?.on(clientEvent: .disconnect) { [weak self] data, _ in
            print("[Socket] Disconnected: \(data)")
            self?.isConnected = false
            let reason = data.first as? String ?? "unknown"
            self?.eventSink?([
                "type": "disconnected",
                "reason": reason
            ])
        }

        socket?.on(clientEvent: .error) { [weak self] data, _ in
            print("[Socket] Error: \(data)")
            let error = data.first as? String ?? "Connection error"
            self?.eventSink?(FlutterError(
                code: "CONNECTION_ERROR",
                message: error,
                details: nil
            ))
        }

        // Protocol event handlers
        setupProtocolEventHandlers()
    }

    private func setupProtocolEventHandlers() {
        // Incoming call
        socket?.on(Protocol.MessageTypes.CALL_INCOMING) { [weak self] data, _ in
            print("[Socket] Received call:incoming")
            if let dict = data.first as? [String: Any] {
                self?.eventSink?([
                    "type": "call:incoming",
                    "data": dict
                ])
            }
        }

        // Call cancelled
        socket?.on(Protocol.MessageTypes.CALL_CANCELLED) { [weak self] data, _ in
            print("[Socket] Received call:cancelled")
            if let dict = data.first as? [String: Any] {
                self?.eventSink?([
                    "type": "call:cancelled",
                    "data": dict
                ])
            }
        }

        // Call ended
        socket?.on(Protocol.MessageTypes.CALL_ENDED) { [weak self] data, _ in
            print("[Socket] Received call:ended")
            if let dict = data.first as? [String: Any] {
                self?.eventSink?([
                    "type": "call:ended",
                    "data": dict
                ])
            }
        }

        // Call failed
        socket?.on(Protocol.MessageTypes.CALL_FAILED) { [weak self] data, _ in
            print("[Socket] Received call:failed")
            if let dict = data.first as? [String: Any] {
                self?.eventSink?([
                    "type": "call:failed",
                    "data": dict
                ])
            }
        }

        // WebRTC offer
        socket?.on(Protocol.MessageTypes.WEBRTC_OFFER) { [weak self] data, _ in
            print("[Socket] Received webrtc:offer")
            if let dict = data.first as? [String: Any] {
                self?.eventSink?([
                    "type": "webrtc:offer",
                    "data": dict
                ])
            }
        }

        // WebRTC ICE candidate
        socket?.on(Protocol.MessageTypes.WEBRTC_ICE_CANDIDATE) { [weak self] data, _ in
            print("[Socket] Received webrtc:ice-candidate")
            if let dict = data.first as? [String: Any] {
                self?.eventSink?([
                    "type": "webrtc:ice-candidate",
                    "data": dict
                ])
            }
        }

        // Device connected
        socket?.on(Protocol.MessageTypes.DEVICE_CONNECTED) { [weak self] data, _ in
            print("[Socket] Received device:connected")
            if let dict = data.first as? [String: Any] {
                self?.eventSink?([
                    "type": "device:connected",
                    "data": dict
                ])
            }
        }

        // Error
        socket?.on(Protocol.MessageTypes.ERROR) { [weak self] data, _ in
            print("[Socket] Received error event")
            if let dict = data.first as? [String: Any] {
                self?.eventSink?([
                    "type": "error",
                    "data": dict
                ])
            }
        }
    }

    private func registerDevice() {
        let deviceId = DeviceUtils.getUniqueDeviceId()

        let deviceConnectEvent: [String: Any] = [
            "type": Protocol.MessageTypes.DEVICE_CONNECT,
            "deviceType": Protocol.DeviceType.mobile.rawValue,
            "deviceId": deviceId,
            "protocolVersion": Protocol.VERSION,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
        ]

        print("[Socket] Registering device with protocol version \(Protocol.VERSION)")
        socket?.emit(Protocol.MessageTypes.DEVICE_CONNECT, deviceConnectEvent)

        // Send initial status as available
        let statusData: [String: Any] = [
            "deviceId": deviceId,
            "status": Protocol.DeviceStatus.available.rawValue,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
        ]
        socket?.emit(Protocol.MessageTypes.DEVICE_STATUS, statusData)
    }

    private func emit(event: String, data: [String: Any]) {
        socket?.emit(event, data)
        print("[Socket] Emitted event: \(event)")
    }

    private func disconnect() {
        socket?.disconnect()
        socket = nil
        manager = nil
        isConnected = false
        print("[Socket] Disconnected and cleared")
    }

    func dispose() {
        disconnect()
        methodChannel.setMethodCallHandler(nil)
    }

    // MARK: - FlutterStreamHandler
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: flutter/android/app/src/main/kotlin/com/callsafe/mobile/socket/SocketChannelHandler.kt
package com.callsafe.mobile.socket

import android.content.Context
import android.util.Log
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.socket.client.IO
import io.socket.client.Socket
import org.json.JSONObject
import com.callsafe.mobile.protocol.Protocol
import com.callsafe.mobile.utils.getUniqueDeviceId

/**
 * Socket.IO platform channel handler
 * Exposes Socket.IO functionality to Flutter
 */
class SocketChannelHandler(
    messenger: BinaryMessenger,
    private val context: Context
) : MethodChannel.MethodCallHandler {
    private val channel = MethodChannel(messenger, CHANNEL_NAME)
    private val eventChannel = EventChannel(messenger, EVENT_CHANNEL_NAME)
    private var eventSink: EventChannel.EventSink? = null

    private var socket: Socket? = null
    private var isConnected = false

    companion object {
        private const val TAG = "SocketChannelHandler"
        private const val CHANNEL_NAME = "com.callsafe.socket"
        private const val EVENT_CHANNEL_NAME = "com.callsafe.socket.events"
    }

    init {
        channel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "connect" -> {
                val token = call.argument<String>("token") ?: return result.error("INVALID_ARGS", "Missing token", null)
                val serverUrl = call.argument<String>("serverUrl") ?: "wss://tunnel.callsafe.tech"
                connect(token, serverUrl, result)
            }
            "disconnect" -> {
                disconnect()
                result.success(null)
            }
            "emit" -> {
                val event = call.argument<String>("event") ?: return result.error("INVALID_ARGS", "Missing event", null)
                val data = call.argument<Map<String, Any>>("data") ?: return result.error("INVALID_ARGS", "Missing data", null)
                emit(event, data)
                result.success(null)
            }
            "isConnected" -> {
                result.success(isConnected)
            }
            else -> result.notImplemented()
        }
    }

    private fun connect(token: String, serverUrl: String, result: MethodChannel.Result) {
        Log.d(TAG, "connect() called with serverUrl: $serverUrl")

        // Disconnect existing connection if any
        disconnect()

        try {
            val options = IO.Options().apply {
                auth = mapOf("token" to token)
                transports = arrayOf("websocket", "polling")
                timeout = 30000
                forceNew = true
            }

            socket = IO.socket(serverUrl, options)

            socket?.apply {
                on(Socket.EVENT_CONNECT) {
                    Log.d(TAG, "Socket connected")
                    isConnected = true
                    eventSink?.success(mapOf(
                        "type" to "connected"
                    ))

                    // Auto-register device
                    registerDevice()
                }

                on(Socket.EVENT_DISCONNECT) { args ->
                    Log.d(TAG, "Socket disconnected: ${args.getOrNull(0)}")
                    isConnected = false
                    eventSink?.success(mapOf(
                        "type" to "disconnected",
                        "reason" to args.getOrNull(0)?.toString()
                    ))
                }

                on(Socket.EVENT_CONNECT_ERROR) { args ->
                    Log.e(TAG, "Socket connection error: ${args.getOrNull(0)}")
                    eventSink?.error("CONNECTION_ERROR", args.getOrNull(0)?.toString(), null)
                }

                // Listen for protocol events
                setupProtocolEventHandlers()

                connect()
                result.success(null)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create socket", e)
            result.error("CONNECTION_FAILED", e.message, null)
        }
    }

    private fun setupProtocolEventHandlers() {
        socket?.apply {
            // Incoming call
            on(Protocol.MessageTypes.CALL_INCOMING) { args ->
                Log.d(TAG, "Received ${Protocol.MessageTypes.CALL_INCOMING}")
                val data = args.getOrNull(0) as? JSONObject
                data?.let {
                    eventSink?.success(mapOf(
                        "type" to "call:incoming",
                        "data" to jsonToMap(it)
                    ))
                }
            }

            // Call cancelled
            on(Protocol.MessageTypes.CALL_CANCELLED) { args ->
                Log.d(TAG, "Received ${Protocol.MessageTypes.CALL_CANCELLED}")
                val data = args.getOrNull(0) as? JSONObject
                data?.let {
                    eventSink?.success(mapOf(
                        "type" to "call:cancelled",
                        "data" to jsonToMap(it)
                    ))
                }
            }

            // Call ended
            on(Protocol.MessageTypes.CALL_ENDED) { args ->
                Log.d(TAG, "Received ${Protocol.MessageTypes.CALL_ENDED}")
                val data = args.getOrNull(0) as? JSONObject
                data?.let {
                    eventSink?.success(mapOf(
                        "type" to "call:ended",
                        "data" to jsonToMap(it)
                    ))
                }
            }

            // Call failed
            on(Protocol.MessageTypes.CALL_FAILED) { args ->
                Log.d(TAG, "Received ${Protocol.MessageTypes.CALL_FAILED}")
                val data = args.getOrNull(0) as? JSONObject
                data?.let {
                    eventSink?.success(mapOf(
                        "type" to "call:failed",
                        "data" to jsonToMap(it)
                    ))
                }
            }

            // WebRTC offer
            on(Protocol.MessageTypes.WEBRTC_OFFER) { args ->
                Log.d(TAG, "Received ${Protocol.MessageTypes.WEBRTC_OFFER}")
                val data = args.getOrNull(0) as? JSONObject
                data?.let {
                    eventSink?.success(mapOf(
                        "type" to "webrtc:offer",
                        "data" to jsonToMap(it)
                    ))
                }
            }

            // WebRTC ICE candidate
            on(Protocol.MessageTypes.WEBRTC_ICE_CANDIDATE) { args ->
                Log.d(TAG, "Received ${Protocol.MessageTypes.WEBRTC_ICE_CANDIDATE}")
                val data = args.getOrNull(0) as? JSONObject
                data?.let {
                    eventSink?.success(mapOf(
                        "type" to "webrtc:ice-candidate",
                        "data" to jsonToMap(it)
                    ))
                }
            }

            // Device connected
            on(Protocol.MessageTypes.DEVICE_CONNECTED) { args ->
                Log.d(TAG, "Received ${Protocol.MessageTypes.DEVICE_CONNECTED}")
                val data = args.getOrNull(0) as? JSONObject
                data?.let {
                    eventSink?.success(mapOf(
                        "type" to "device:connected",
                        "data" to jsonToMap(it)
                    ))
                }
            }

            // Error
            on(Protocol.MessageTypes.ERROR) { args ->
                Log.e(TAG, "Received error event")
                val error = args.getOrNull(0) as? JSONObject
                error?.let {
                    eventSink?.success(mapOf(
                        "type" to "error",
                        "data" to jsonToMap(it)
                    ))
                }
            }
        }
    }

    private fun registerDevice() {
        val deviceId = getUniqueDeviceId(context)

        val deviceConnectEvent = JSONObject().apply {
            put("type", Protocol.MessageTypes.DEVICE_CONNECT)
            put("deviceType", Protocol.DeviceType.MOBILE.value)
            put("deviceId", deviceId)
            put("protocolVersion", Protocol.VERSION)
            put("timestamp", System.currentTimeMillis())
        }

        Log.d(TAG, "Registering device with protocol version ${Protocol.VERSION}")
        socket?.emit(Protocol.MessageTypes.DEVICE_CONNECT, deviceConnectEvent)

        // Send initial status as available
        val statusData = JSONObject().apply {
            put("deviceId", deviceId)
            put("status", Protocol.DeviceStatus.AVAILABLE.value)
            put("timestamp", System.currentTimeMillis())
        }
        socket?.emit(Protocol.MessageTypes.DEVICE_STATUS, statusData)
    }

    private fun emit(event: String, data: Map<String, Any>) {
        val jsonData = JSONObject(data)
        socket?.emit(event, jsonData)
        Log.d(TAG, "Emitted event: $event")
    }

    private fun disconnect() {
        socket?.disconnect()
        socket = null
        isConnected = false
        Log.d(TAG, "Socket disconnected and cleared")
    }

    private fun jsonToMap(json: JSONObject): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        json.keys().forEach { key ->
            val value = json.get(key)
            map[key] = when (value) {
                is JSONObject -> jsonToMap(value)
                JSONObject.NULL -> null
                else -> value
            }
        }
        return map
    }

    fun dispose() {
        disconnect()
        channel.setMethodCallHandler(null)
    }
}
